src/lib/operators/index_scan.cpp

```diff
@@ -6,203 +6,142 @@
-      range_begin = index->cbegin();
-      range_end = index->upper_bound(_right_values);
-      break;
-    }
-    case PredicateCondition::GreaterThan: {
-      range_begin = index->upper_bound(_right_values);
-      range_end = index->cend();
-      break;
-    }
-    case PredicateCondition::GreaterThanEquals: {
-      range_begin = index->lower_bound(_right_values);
-      range_end = index->cend();
-      break;
-    }
-    case PredicateCondition::BetweenInclusive: {
-      range_begin = index->lower_bound(_right_values);
-      range_end = index->upper_bound(_right_values2);
-      break;
-    }
-    case PredicateCondition::BetweenLowerExclusive: {
-      range_begin = index->upper_bound(_right_values);
-      range_end = index->upper_bound(_right_values2);
-      break;
-    }
-    case PredicateCondition::BetweenUpperExclusive: {
-      range_begin = index->lower_bound(_right_values);
-      range_end = index->lower_bound(_right_values2);
-      break;
-    }
-    case PredicateCondition::BetweenExclusive: {
-      range_begin = index->upper_bound(_right_values);
-      range_end = index->lower_bound(_right_values2);
+      index->range_not_equals_with_iterators(append_matches, _scan_value);
       break;
     }
     default:
-      Fail("Unsupported comparison type encountered");
+      Fail("Unsupported comparison type. Currently, Hyrise's secondary indexes only support Equals and NotEquals.");
+  }
+
+  if (pos_lists[0]->empty()) {
+    return _out_table;
   }
 
-  DebugAssert(_in_table->type() == TableType::Data, "Cannot guarantee single chunk PosList for non-data tables.");
-  matches_out.guarantee_single_chunk();
+  const auto in_table_column_count = _in_table->column_count();
+  for (const auto& pos_list : pos_lists) {
+    auto segments = Segments{};
+    segments.reserve(in_table_column_count);
 
-  const auto current_matches_size = matches_out.size();
-  const auto final_matches_size = current_matches_size + static_cast<size_t>(std::distance(range_begin, range_end));
-  matches_out.resize(final_matches_size);
+    for (auto column_id = ColumnID{0}; column_id < in_table_column_count; ++column_id) {
+      segments.emplace_back(std::make_shared<ReferenceSegment>(_in_table, column_id, pos_list));
+    }
 
-  for (auto matches_position = current_matches_size; matches_position < final_matches_size; ++matches_position) {
-    matches_out[matches_position] = RowID{chunk_id, *range_begin};
-    ++range_begin;
+    _out_table->append_chunk(segments, nullptr);
   }
 
-  return matches_out;
+  return _out_table;
+}
+
+std::shared_ptr<AbstractOperator> IndexScan::_on_deep_copy(
+    const std::shared_ptr<AbstractOperator>& copied_left_input,
+    const std::shared_ptr<AbstractOperator>& /*copied_right_input*/,
+    std::unordered_map<const AbstractOperator*, std::shared_ptr<AbstractOperator>>& /*copied_ops*/) const {
+  return std::make_shared<IndexScan>(copied_left_input, _indexed_column_id, _predicate_condition, _scan_value);
```

Comment:
It's changed in the `_on_execute()` method. Do we copy executed operators somewhere?

But anyhow, I see your point. I think the easy solution is to have a local var instead of modifying the member variable. What do you think?

Reply:
(so, keep the member, but `_on_execute()` stores the "new" column id in a local variable.

Reply:
Ahh, the PQP cache copies. Right ...